cli/rt/21_filereader.js

```diff
@@ -0,0 +1,206 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+
+((window) => {
+  const base64 = window.__bootstrap.base64;
+
+  async function readOperation(fr, blob, readtype) {
+    // Implementation from https://w3c.github.io/FileAPI/ notes
+    // And body of deno blob.ts readBytes
+
+    // 1. If fr’s state is "loading", throw an InvalidStateError DOMException.
+    if (fr.readyState === FileReader.LOADING) {
+      throw new DOMExcechunkPromiseption(
+        "Invalid FileReader state.",
+        "InvalidStateError",
+      );
+    }
+    // 2. Set fr’s state to "loading".
+    fr.readyState = FileReader.LOADING;
+    // 3. Set fr’s result to null.
+    fr.result = null;
+    // 4. Set fr’s error to null.
+    fr.error = null;
+
+    // 5. Let stream be the result of calling get stream on blob.
+    const stream /*: ReadableStream<ArrayBufferView>*/ = blob.stream();
+
+    // 6. Let reader be the result of getting a reader from stream.
+    const reader = stream.getReader();
+
+    // 7. Let bytes be an empty byte sequence.
+    //let bytes = new Uint8Array();
+    const chunks /*: Uint8Array[]*/ = [];
+
+    // 8. Let chunkPromise be the result of reading a chunk from stream with reader.
+    let chunkPromise = reader.read();
+
+    // 9. Let isFirstChunk be true.
+    let isFirstChunk = true;
+
+    // 10 in parallel while true
+    while (true) {
+      // 1. Wait for chunkPromise to be fulfilled or rejected.
+      try {
+        const chunk = await chunkPromise;
+
+        // 2. If chunkPromise is fulfilled, and isFirstChunk is true, queue a task to fire a progress event called loadstart at fr.
+        if (isFirstChunk) {
+          setTimeout(() => {
+            // fire a progress event for loadstart
+            const ev = new ProgressEvent("loadstart", {});
```

Comment:
Ditto but for `02_progress_event.js`, I suggest putting it in this file instead

Reply:
I put `ProgressEvent` in its own .js file because it is a separate class than FileReader and used in multiple places in dom and webworker .d.ts files already.

Reply:
Okay, but it's not used in runtime anywhere besides `21_filereader.js`. If we add more APIs using `ProgressEvent` then I'll be happy to move it into its own file.